README.md

```diff
@@ -1,78 +1,406 @@
+
+If you open up `output.csv` you'll see the following:
+
+~~~
+firstName
+15-09-1971 04:30:18
+28-10-2087 09:23:30
+1876153676
+334227637
+
+03-04-2053 04:57:06
+"Lorem Ipsum"
+[...]
+~~~
+
+The generator has successfully created 100 rows of random data. However, in most cases this data will not be terribly useful. It is likely that `firstName` will only allow string values, with this constraint enforced via a database schema for example. If you don't provide any constraints, the generator will output values from the 'universal set', which  contains all generatable values (null, any string, any date, any number, etc).
+
+Let's assume you only want to generate string values for the `firstName` field; this can be achieved by adding an `ofType` constraint for the field. With this constraint alone, the generator will output random strings containing characters from the unicode [Basic Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane). If you want a more restrictive character set, ths can be achieved by adding `matchingRegex`. 
+
+
+## Adding constraints
+
+Update the JSON profile as follows:
+
+~~~json
+{
+  "schemaVersion": "0.1",
+  "fields": [
+    { "name": "firstName" }
+  ],
+  "rules": [
+    {
+      "rule": "first name constraints",
+      "constraints": [
+        { "field": "firstName", "is": "ofType", "value": "string" },
+        { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" }
+      ]
+    }
+  ]
+}
+~~~
+
+Re-running generation now creates a file containing random strings that match the simple regex `[a-z]{1,10}`:
+
+~~~
+firstName
+"f"
+
+"draxonsa"
+"j"
+"srsub"
+
+"f"
+
+[...]
+~~~
+
+Fields are nullable by default, however, you can add a further constraint to a field to indicate it is 'not null'. If you add the following constraint, the generator will no longer output null values:
+
+~~~json
+"constraints": [
+  { "field": "firstName", "is": "ofType", "value": "string" },
+  { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" },
+  { "not": { "field": "firstName", "is": "null" } }
+]
+~~~
+
+The generator supports three different types of constraint. These are:
+
+ - **Predicates** - boolean-valued functions that define whether a given value as valid or invalid. In the above profile `ofType` and `matchingRegex` are examples of predicates.
+ - **Grammatical** - combine or modify other constraints. They are fully recursive; any grammatical constraint is a valid input to any other grammatical constraint. In the above profile `not` is an example of a grammatical constraint. 
+ - **Presentational** - provide additional formatting information, we'll cover these later.
```

Comment:
Can anyone think of a good example for this? I haven't actually covered them.

Reply:
- Formatting? 
- Output? 
- Serialisation?

Formatting seems the best, but maybe thats because we only have one constraint in this group, the `formattedAs` constraint!

Reply:
Ah - I actually meant a specific example that made sense in the context of this getting started guide!

Reply:
I cant think of a good example with the fields you have, presentation constraints are best performed with datetimes or with larger (or high precision) numbers.

Reply:
OK, perhaps leave this for now.